power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -273,41 +284,598 @@ inline auto rank_transformers(State const& state) -> RankedTransformerGroups {
+    using RegulatedTransformer = TapRegulatorRef<TransformerTypes...>;
+    using UpdateBuffer = std::tuple<std::vector<typename TransformerTypes::UpdateType>...>;
+
+    template <transformer_c T>
+    static constexpr auto transformer_index_of = container_impl::get_cls_pos_v<T, TransformerTypes...>;
+    static_assert(((transformer_index_of<TransformerTypes> < sizeof...(TransformerTypes)) && ...));
 
-    TapPositionOptimizer(Calculator calculator, StateUpdater updater, OptimizerStrategy strategy)
+  public:
+    TapPositionOptimizerImpl(Calculator calculator, StateUpdater updater, OptimizerStrategy strategy)
         : calculate_{std::move(calculator)}, update_{std::move(updater)}, strategy_{strategy} {}
 
-    auto optimize(State const& state) -> ResultType final {
-        auto const order = rank_transformers(state);
-        return optimize(state, order);
+    auto optimize(State const& state, CalculationMethod method) -> ResultType final {
+        auto const order = regulator_mapping<TransformerTypes...>(state, TransformerRanker{}(state));
+
+        auto const cache = this->cache_states(order);
+        auto optimal_output = optimize(state, order, method);
+        update_state(cache);
+
+        return optimal_output;
     }
 
-    constexpr auto get_strategy() { return strategy_; }
+    constexpr auto get_strategy() const { return strategy_; }
 
   private:
-    auto optimize(State const& /*state*/, RankedTransformerGroups const& /*order*/) -> ResultType {
-        // TODO(mgovers): implement outter loop tap changer
-        throw PowerGridError{};
+    auto optimize(State const& state, std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                  CalculationMethod method) const -> ResultType {
+        initialize(regulator_order);
+
+        if (auto result = iterate(state, regulator_order, method); strategy_ == OptimizerStrategy::any) {
+            return result;
+        }
+
+        // refine solution
+        step_all_back(regulator_order);
+        return iterate(state, regulator_order, method);
+    }
+
+    auto iterate(State const& state, std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                 CalculationMethod method) const -> ResultType {
+        auto result = calculate_with_retry(state, method);
+
+        bool tap_changed = true;
+        while (tap_changed) {
+            tap_changed = false;
+            UpdateBuffer update_data;
+
+            for (auto const& same_rank_regulators : regulator_order) {
+                for (auto const& regulator : same_rank_regulators) {
+                    tap_changed = tap_changed || control_transformer(regulator, state, result, update_data);
+                }
+                if (tap_changed) {
+                    break;
+                }
+            }
+            if (tap_changed) {
+                update_state(update_data);
+                result = calculate_(state, method);
```

Comment:
I have one doubt here: is it guaranteed that `calculate_(state, method)` is good? No need to specify other `::linear` alternative?

```c++
// code changed somethere else
auto calculate_with_retry(State const& state, CalculationMethod method_in, CalculationMethod & method_used) const {
    try {
        method_used = method_in;
        return calculate_(state, method);
    } catch (const SparseMatrixError& e, const IterationDiverge& e) {
        method_used = CalculationMethod::linear;
        return calculate_(state, CalculationMethod::linear);
    }
}
```
and then:
```c++
    auto iterate(State const& state, std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
                 CalculationMethod method) const -> ResultType {
        CalculationMethod method_used;
        auto result = calculate_with_retry(state, method, method_used);

        bool tap_changed = true;
        while (tap_changed) {
            tap_changed = false;
            UpdateBuffer update_data;

            for (auto const& same_rank_regulators : regulator_order) {
                for (auto const& regulator : same_rank_regulators) {
                    tap_changed = tap_changed || control_transformer(regulator, state, result, update_data);
                }
                if (tap_changed) {
                    break;
                }
            }
            if (tap_changed) {
                update_state(update_data);
                result = calculate_(state, method_used);
            }
        }

        add_tap_regulator_output(state, regulator_order, result);

        return result;
    }
```

Reply:
> I have one doubt here: is it guaranteed that calculate_(state, method) is good? No need to specify other ::linear alternative?

indeed. in fact, I believe that with #505 in place, this might not even be necessary anymore. @TonyXiang8787 do you know?

Reply:
```cpp
CalculationMethod method_used;
auto result = calculate_with_retry(state, method, method_used);
// ...
result = calculate_(state, method_used);
```

this would mean it will always use `linear` as a fallback and keep using it, while in practice, it should only be an improved initial guess to ensure convergence on the first run that is later refined with the actual provided method

Reply:
`calculate_with_retry` says otherwise

Reply:
no, the current `calculate_with_retry` is only used during the initial result. The fallback `linear` method is used there. During the actual calculation, the `method` provided by the user is used.

In your suggestion, you set `method_used` based on which method you provided and you only use that, completely ignoring whatever the user provided.

Reply:
i have a proposal: let's remove the `calculate_with_retry` and only use `calculate_`. The retry logic is something that can be worked out in a follow-up step. This suits the Continuous Improvement WoW. Since @TonyXiang8787 's input is needed here anyways, I would deem this a next iteration

Reply:
> no, the current `calculate_with_retry` is only used during the initial result. The fallback `linear` method is used there. During the actual calculation, the `method` provided by the user is used.

`calculate_with_retry` is called every time `iterate` is called.

> In your suggestion, you set `method_used` based on which method you provided and you only use that, completely ignoring whatever the user provided.

That's what the code in its original state suggests: if the users method doesn't work, use linear. Twice.

Reply:
OK, i do see a mistake in my implementation:
```
try 
  math_result = math_solver.power_flow(calculation_method=calculation_method) 
  if (abs(u_pu) < 0.2).any(): 
    Throw InvalidPFResult 
catch (InvalidPFResult, SparseMatrixError, IterationDiverge) 
  Run run_pf_with_regulation_per_solver(math_solver=math_solver, calculation_method=linear),   discard return 
  math_result = math_solver.power_flow(calculation_method=calculation_method) // i forgot this line
```

I added a Jira ticket for that because I have reason to believe that #505 should solve this because of the discard step: the initial state is actually set to the solution of the `linear` calculation.

That leaves me with even more reason that all the other fallback logic is also unnecessary and we can just remove `calculate_with_retry` and, instead, directly call `calculate_`

Reply:
removing from this PR.

Reply:
Moved to https://github.com/PowerGridModel/power-grid-model/issues/580